File: web/src/engine/signals.ts
/**
 * 可观测信号抽取（对应 Python `gates/signals.py`）。
 *
 * 这是"防自证"的第一道物理隔离：门禁不直接摸达人原始对象，而是先过这份白名单。
 * 想读 ground truth 也读不到，且标定与判定共用同一个抽取函数，
 * 不可能出现"标定用 A 口径、判定用 B 口径"的经典 bug。
 */

import type { GatePolicy } from './policy.ts';
import { safeDiv } from './stats.ts';
import { followerBucket, groupKey } from './taxonomy.ts';
import type { Kox } from './types.ts';

/** 参与分位数标定的信号名（与 thresholds.json 的信号 key 一致）。 */
export const OBSERVABLE_SIGNALS = [
  'engagement_rate',
  'comment_like_ratio',
  'view_follower_ratio',
  'followers_per_day',
  'comment_dup_rate',
  'comment_emoji_only_rate',
  'max_monthly_growth',
  'risk_severity_score',
  'avg_cpm_usd',
] as const;

export type SignalName = (typeof OBSERVABLE_SIGNALS)[number];

/** 安全取数：null / 非数值 / 布尔一律视为缺失（与 Python `_num` 同语义）。 */
export function numOrNull(value: unknown): number | null {
  if (value === null || value === undefined || typeof value === 'boolean') return null;
  if (typeof value === 'number') return Number.isFinite(value) ? value : null;
  return null;
}

/** 分位数分组键 `platform|bucket`；followers 缺失时落入 nano 组（最保守）。 */
export function koxGroupKey(kox: Partial<Kox>): string {
  return groupKey(String(kox.platform ?? 'unknown'), followerBucket(numOrNull(kox.followers)));
}

/**
 * 累计风险载荷分（只统计 low/medium；high 走 G3.1 硬阻断，不参与累加）。
 * 单条 flag 的 hits 在 `severity_hits_cap` 处截断：命中 10 次和 3 次对决策含义差别不大，
 * 不截断会让分数被单条 flag 主导。
 */
export function riskSeverityScore(kox: Partial<Kox>, policy: GatePolicy): number {
  let total = 0.0;
  for (const flag of kox.content_flags ?? []) {
    if (!flag || typeof flag !== 'object') continue;
    const severity = String(flag.severity ?? 'low');
    if (severity === 'high') continue;
    const hits = numOrNull(flag.hits) ?? 1.0;
    total += (policy.severity_points[severity] ?? 1.0) * Math.min(hits, policy.severity_hits_cap);
  }
  const controversy = kox.controversy;
  if (controversy && typeof controversy === 'object') {
    total += policy.severity_points[String(controversy.severity ?? 'low')] ?? 1.0;
  }
  return total;
}

/** 粉丝历史 -> 月度环比增速序列（长度 = len(history) - 1）。 */
export function monthlyGrowthRates(history: readonly unknown[] | null | undefined): number[] {
  if (!history || history.length < 2) return [];
  const rates: number[] = [];
  for (let i = 0; i + 1 < history.length; i += 1) {
    const p = numOrNull(history[i]);
    const c = numOrNull(history[i + 1]);
    if (p === null || c === null || p <= 0) rates.push(0.0);
    else rates.push((c - p) / p);
  }
  return rates;
}

export type SignalBag = Record<SignalName, number | null>;

/**
 * 抽出全部可观测标量信号。缺失一律 null，绝不用 0 冒充
 * （否则 P05 阈值会被一堆假 0 拉到地板上）。
 */
export function extractSignals(kox: Partial<Kox>, policy: GatePolicy): SignalBag {
  const followers = numOrNull(kox.followers);
  const ageDays = numOrNull(kox.account_age_days);
  let fpd: number | null = null;
  if (followers !== null && ageDays !== null && ageDays > 0) fpd = safeDiv(followers, ageDays);
  const growth = monthlyGrowthRates(kox.follower_history);
  return {
    engagement_rate: numOrNull(kox.engagement_rate),
    comment_like_ratio: numOrNull(kox.comment_like_ratio),
    view_follower_ratio: numOrNull(kox.view_follower_ratio),
    followers_per_day: fpd,
    comment_dup_rate: numOrNull(kox.comment_dup_rate),
    comment_emoji_only_rate: numOrNull(kox.comment_emoji_only_rate),
    max_monthly_growth: growth.length > 0 ? Math.max(...growth) : null,
    risk_severity_score: riskSeverityScore(kox, policy),
    avg_cpm_usd: numOrNull(kox.avg_cpm_usd),
  };
}

File: src/koxpilot/gates/signals.py
"""可观测信号抽取：门禁引擎能看到的全部输入，就是这里列出的东西。

设计意图（防自证的第一道物理隔离）
------------------------------------
门禁不直接摸 kox 原始 dict，而是先经过本模块把"可观测信号"抽出来。
好处有三：
1. 想读 ground truth 也读不到 —— 抽取函数只认这份白名单；
2. 阈值标定与规则判定用**同一个抽取函数**，不可能出现"标定用 A 口径、判定用 B 口径"的经典 bug；
3. 缺失字段（``_missing``）在这里统一变成 ``None``，各规则只需判断 None 即可跳过。
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any, Final

from ..stats import safe_div
from ..taxonomy import follower_bucket, group_key
from .policy import SEVERITY_HITS_CAP, SEVERITY_POINTS

__all__ = [
    "OBSERVABLE_SIGNALS",
    "extract_signals",
    "kox_group_key",
    "monthly_growth_rates",
    "risk_severity_score",
]

#: 参与分位数标定的信号名（与 policy.SIGNAL_QUANTILES 的 key 一致）
OBSERVABLE_SIGNALS: Final[tuple[str, ...]] = (
    "engagement_rate",
    "comment_like_ratio",
    "view_follower_ratio",
    "followers_per_day",
    "comment_dup_rate",
    "comment_emoji_only_rate",
    "max_monthly_growth",
    "risk_severity_score",
    "avg_cpm_usd",
)


def _num(value: Any) -> float | None:
    """安全取数：None / 非数值 / 布尔 一律视为缺失。"""
    if value is None or isinstance(value, bool):
        return None
    if isinstance(value, (int, float)):
        return float(value)
    return None


def kox_group_key(kox: Mapping[str, Any]) -> str:
    """分位数分组键 ``platform|bucket``。followers 缺失时落入 nano 组（最保守）。"""
    return group_key(str(kox.get("platform", "unknown")), follower_bucket(_num(kox.get("followers"))))


def risk_severity_score(kox: Mapping[str, Any]) -> float:
    """累计风险载荷分（只统计 low/medium；high 走 G3.1 硬阻断，不参与累加）。

    单条 flag 的 hits 在 SEVERITY_HITS_CAP 处截断：一条 flag 命中 10 次和命中 3 次，
    对品牌方的决策含义差别不大，但不截断会让分数被单条 flag 主导。
    """
    total = 0.0
    for flag in kox.get("content_flags") or []:
        if not isinstance(flag, Mapping):
            continue
        severity = str(flag.get("severity", "low"))
        if severity == "high":
            continue
        hits = _num(flag.get("hits")) or 1.0
        total += SEVERITY_POINTS.get(severity, 1.0) * min(hits, float(SEVERITY_HITS_CAP))
    controversy = kox.get("controversy")
    if isinstance(controversy, Mapping):
        total += SEVERITY_POINTS.get(str(controversy.get("severity", "low")), 1.0)
    return total


def monthly_growth_rates(history: list[Any] | None) -> list[float]:
    """粉丝历史 -> 月度环比增速序列（长度 = len(history) - 1）。"""
    if not history or len(history) < 2:
        return []
    rates: list[float] = []
    for prev, cur in zip(history, history[1:]):
        p, c = _num(prev), _num(cur)
        if p is None or c is None or p <= 0:
            rates.append(0.0)
        else:
            rates.append((c - p) / p)
    return rates


def extract_signals(kox: Mapping[str, Any]) -> dict[str, float | None]:
    """抽出全部可观测标量信号。缺失一律 None，绝不用 0 冒充。

    ``followers_per_day`` 是派生信号（G1.6「新号巨量粉」），
    在这里派生而不是在规则里派生，是为了让标定与判定共享同一定义。
    """
    followers = _num(kox.get("followers"))
    age_days = _num(kox.get("account_age_days"))
    fpd: float | None = None
    if followers is not None and age_days is not None and age_days > 0:
        fpd = safe_div(followers, age_days)
    growth = monthly_growth_rates(kox.get("follower_history"))
    return {
        "engagement_rate": _num(kox.get("engagement_rate")),
        "comment_like_ratio": _num(kox.get("comment_like_ratio")),
        "view_follower_ratio": _num(kox.get("view_follower_ratio")),
        "followers_per_day": fpd,
        "comment_dup_rate": _num(kox.get("comment_dup_rate")),
        "comment_emoji_only_rate": _num(kox.get("comment_emoji_only_rate")),
        "max_monthly_growth": max(growth) if growth else None,
        "risk_severity_score": risk_severity_score(kox),
        "avg_cpm_usd": _num(kox.get("avg_cpm_usd")),
    }
